crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,387 @@
+        }
+    }
+
+    /// Returns the next pipeline state depending on the current value of the next action.
+    /// Resets the next action to the default value.
+    fn next_pipeline_state(
+        &mut self,
+        pipeline: Pipeline<DB, U>,
+        tip: H256,
+    ) -> PipelineState<DB, U> {
+        let next_action = std::mem::take(&mut self.next_action);
+        if next_action.run_pipeline() {
+            trace!(target: "sync::controller", ?tip, "Starting the pipeline");
+            PipelineState::Running(pipeline.run_as_fut(self.db.clone(), tip))
+        } else {
+            PipelineState::Idle(pipeline)
+        }
+    }
+
+    /// Attempt to restore the tree with the finalized block number.
+    /// If the finalized block is missing from the database, trigger the pipeline run.
+    fn restore_tree_if_possible(
+        &mut self,
+        finalized_hash: BlockHash,
+    ) -> Result<(), BeaconEngineError> {
+        match self.db.view(|tx| tx.get::<tables::HeaderNumbers>(finalized_hash))?? {
+            Some(number) => self.blockchain_tree.restore_canonical_hashes(number)?,
+            None => self.pipeline_run_needed(),
+        };
+        Ok(())
+    }
+}
+
+/// On initialization, the controller will poll the message receiver and return [Poll::Pending]
+/// until the first forkchoice update message is received.
+///
+/// As soon as the controller receives the first forkchoice updated message and updates the local
+/// forkchoice state, it will launch the pipeline to sync to the head hash.
+/// While the pipeline is syncing, the controller will keep processing messages from the receiver
+/// and forwarding them to the blockchain tree.
+impl<DB, U, C, EF> Future for BeaconConsensusEngine<DB, U, C, EF>
+where
+    DB: Database + Unpin + 'static,
+    U: SyncStateUpdater + Unpin + 'static,
+    C: Consensus + Unpin,
+    EF: ExecutorFactory + Unpin + 'static,
+{
+    type Output = Result<(), BeaconEngineError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        // Process all incoming messages first.
+        while let Poll::Ready(Some(msg)) = this.message_rx.poll_next_unpin(cx) {
+            match msg {
+                BeaconEngineMessage::ForkchoiceUpdated(state, tx) => {
+                    let response = this.on_forkchoice_updated(state);
+                    let _ = tx.send(response);
+                }
+                BeaconEngineMessage::NewPayload(block, tx) => {
+                    let response = this.on_new_payload(block);
+                    let _ = tx.send(response);
+                }
+            }
+        }
+
+        // Set the next pipeline state.
+        loop {
+            // Lookup the forkchoice state. We can't launch the pipeline without the tip.
+            let forckchoice_state = match &this.forkchoice_state {
+                Some(state) => state,
+                None => return Poll::Pending,
```

Comment:
are we missing any ctx waker wake by ref's here / any budget? @mattsse

Reply:
I don't think so, the state will always be None, right?
so can't react to any outside changes besides new messages, which are already drained.

Reply:
what @mattsse said